src/pgwire/src/protocol.rs

```diff
@@ -139,6 +147,13 @@ where
             None => "eof",
         };
 
+        // Clear any cancellation message.
+        // TODO(mjibson): This makes the use of .changed annoying since it will
+        // generally always have a NotCancelled message first that needs to be ignored,
+        // and thus run in a loop. Figure out a way to have the future only resolve on
+        // a Cancelled message.
+        let _ = self.cancel_tx.lock().unwrap().send(Cancelled::NotCancelled);
```

Comment:
If you're on board with moving this to `SessionClient` (see below), then I think there is a really nice place for this logic:

```rust
impl SessionClient {
    async fn canceled(&self) {
        loop {
            self.cancel_rx.changed().await;
            if let Cancelled::Cancelled = *cancel.borrow() {
                return;
            }
        }
    }
}
```